File: app/src/main/java/me/sunzheng/mana/FavoriteCompact.java
package me.sunzheng.mana;

import android.annotation.TargetApi;
import android.content.Context;
import android.content.res.ColorStateList;
import android.graphics.drawable.ColorDrawable;
import android.graphics.drawable.Drawable;
import android.graphics.drawable.RippleDrawable;
import android.os.Build;
import android.widget.TextView;

import androidx.core.content.ContextCompat;

/**
 * Created by Sun on 2017/7/6.
 */

public class FavoriteCompact {
    static int[] colors = new int[]{android.R.attr.buttonStyle, R.color.favorite_wish, R.color.favorite_watched, R.color.favorite_watching, R.color.favorite_pause, R.color.favorite_abanoned};

    public static void setFavorite(long status, TextView v) {
        String statusString = v.getResources().getStringArray(R.array.favorite_status_values)[(int) status];
        DrawableCompat drawableCompat = null;
        if (Build.VERSION.SDK_INT >= 21)
            drawableCompat = new LolipopDrawableCompatImpl(v.getContext());
        else
            drawableCompat = new BaseDrawableCompatImpl(v.getContext());
//        v.setBackground(drawableCompat.getDrawable(status,v));
//        switch ((int) status) {
//            case 1:
//                v.setBackgroundColor(ContextCompat.getColor(v.getContext(), R.color.favorite_wish));
//                break;
//            case 2:
//                v.setBackgroundColor(ContextCompat.getColor(v.getContext(), R.color.favorite_watched));
//                break;
//            case 3:
//                v.setBackgroundColor(ContextCompat.getColor(v.getContext(), R.color.favorite_watching));
//                break;
//            case 4:
//                v.setBackgroundColor(ContextCompat.getColor(v.getContext(), R.color.favorite_pause));
//                break;
//            case 5:
//                v.setBackgroundColor(ContextCompat.getColor(v.getContext(), R.color.favorite_abanoned));
//                break;
//            default:
//                break;
//        }
        v.setText(statusString);
    }

    static abstract class DrawableCompat {
        private Context mContext;

        public DrawableCompat(Context mContext) {
            this.mContext = mContext;
        }

        public Context getContext() {
            return mContext;
        }

        abstract Drawable getDrawable(long status, TextView v);
    }

    static class BaseDrawableCompatImpl extends DrawableCompat {
        public BaseDrawableCompatImpl(Context mContext) {
            super(mContext);
        }

        @Override
        Drawable getDrawable(long status, TextView v) {
            return null;
        }
    }

    @TargetApi(21)
    static class LolipopDrawableCompatImpl extends DrawableCompat {
        public LolipopDrawableCompatImpl(Context mContext) {
            super(mContext);
        }

        @Override
        Drawable getDrawable(long status, TextView v) {
            int[][] cstatus = new int[][]{new int[]{android.R.attr.state_pressed}};
            return new RippleDrawable(new ColorStateList(cstatus, colors),
                    new ColorDrawable(ContextCompat.getColor(v.getContext(), colors[(int) status % colors.length])),
                    v.getBackground());
        }
    }

}

File: app/src/main/java/me/sunzheng/mana/FavoriteCompact.kt
package me.sunzheng.mana

import android.annotation.TargetApi
import android.content.Context
import android.content.res.ColorStateList
import android.graphics.drawable.ColorDrawable
import android.graphics.drawable.Drawable
import android.graphics.drawable.RippleDrawable
import android.widget.TextView
import androidx.core.content.ContextCompat

/**
 * Created by Sun on 2017/7/6.
 */
object FavoriteCompact {
    var colors = intArrayOf(
        android.R.attr.buttonStyle,
        R.color.favorite_wish,
        R.color.favorite_watched,
        R.color.favorite_watching,
        R.color.favorite_pause,
        R.color.favorite_abanoned
    )

    fun setFavorite(status: Int, v: TextView) {
        val statusString =
            v.resources.getStringArray(R.array.favorite_status_values)[status.toInt()]
        var colorInt = when (status) {
            1 -> R.color.favorite_wish
            2 -> R.color.favorite_watched
            3 -> R.color.favorite_watching
            4 -> R.color.favorite_pause
            5 -> R.color.favorite_abanoned
            else -> R.color.favorite_abanoned
        }
        v.setBackgroundColor(ContextCompat.getColor(v.context, colorInt))
        v.text = statusString
    }

    internal abstract class DrawableCompat(val context: Context) {
        abstract fun getDrawable(status: Long, v: TextView): Drawable?
    }

    internal class BaseDrawableCompatImpl(mContext: Context) : DrawableCompat(mContext) {
        override fun getDrawable(status: Long, v: TextView): Drawable? {
            return null
        }
    }

    @TargetApi(21)
    internal class LolipopDrawableCompatImpl(mContext: Context) : DrawableCompat(mContext) {
        override fun getDrawable(status: Long, v: TextView): Drawable? {
            val cstatus = arrayOf(intArrayOf(android.R.attr.state_pressed))
            return RippleDrawable(
                ColorStateList(cstatus, colors),
                ColorDrawable(
                    ContextCompat.getColor(
                        v.context,
                        colors[status.toInt() % colors.size]
                    )
                ),
                v.background
            )
        }
    }
}